elephant/causality/granger.py

```diff
@@ -0,0 +1,382 @@
+
+    C_0 = C[0] - \sum_{i=0}^{p-1} A[i]C[i+1]
+
+    where p is the orde of the autoregressive model.
+
+    """
+
+    yule_walker_matrix, lag_covariances = \
+        _yule_walker_matrix(signals, dimension, order)
+
+    positive_lag_covariances = np.reshape(lag_covariances[1:],
+                                          (dimension*order, dimension))
+
+    lstsq_coeffs = \
+        np.linalg.lstsq(yule_walker_matrix, positive_lag_covariances)[0]
+
+    coeffs = []
+    for index in range(order):
+        coeffs.append(lstsq_coeffs[index*dimension:(index+1)*dimension, ].T)
+
+    coeffs = np.stack(coeffs)
+
+    cov_matrix = np.copy(lag_covariances[0])
+    for i in range(order):
+        cov_matrix -= np.matmul(coeffs[i], lag_covariances[i+1])
+
+    return coeffs, cov_matrix
+
+
+def _optimal_vector_arm(signals, dimension, max_order,
+                        information_criterion='bic'):
+    """
+    Determine optimal auto regressive model by choosing optimal order via
+    Information Criterion
+    Parameters
+    ----------
+    signals : np.ndarray
+        time series data
+    dimension : int
+        dimensionality of the data
+    max_order : int
+        maximal order to consider
+    information_criterion : string
+        bic for Bayesian information_criterion,
+        aic for Akaike information criterion
+
+    Returns
+    -------
+    optimal_coeffs: np.ndarray
+        coefficients of the autoregressive model
+    optimal_cov_mat : np.ndarray
+        covariance matrix of
+    optimal_order : int
+        optimal order
+    """
+
+    length = np.size(signals[0])
+
+    optimal_ic = np.infty
+    optimal_order = 1
+    optimal_coeffs = np.zeros((dimension, dimension, optimal_order))
+    optimal_cov_matrix = np.zeros((dimension, dimension))
+
+    if information_criterion == 'bic':
+        evaluate_ic = bic
+    elif information_criterion == 'aic':
+        evaluate_ic = aic
+    else:
+        raise ValueError(
+            f"Information criterion {information_criterion} not valid")
+
+    for order in range(1, max_order + 1):
+        coeffs, cov_matrix = _vector_arm(signals, dimension, order)
+
+        temp_ic = evaluate_ic(cov_matrix, order, dimension, length)
+
+        if temp_ic < optimal_ic:
+            optimal_ic = temp_ic
+            optimal_order = order
+            optimal_coeffs = coeffs
+            optimal_cov_matrix = cov_matrix
+
+    return optimal_coeffs, optimal_cov_matrix, optimal_order
+
+
+def pairwise_granger(signals, max_order, information_criterion='bic'):
+    r"""
+    Determine Granger Causality of two time series
+    Note: order parameter should be removed
+    Parameters
+    ----------
+    signals : np.ndarray or neo.AnalogSignal
+        time series data
+    max_order : int
+        maximal order of autoregressive model
+    information_criterion : string
+        bic for Bayesian information_criterion,
+        aic for Akaike information criterion
+    Returns
```

Comment:
any two doc sections should be separated with a new line
apply to all public functions

Reply:
Addressed in commit daa2bc6